lxd/patches.go

```diff
@@ -1436,4 +1437,41 @@ func patchInstanceRemoveVolatileLastStateIPAddresses(_ string, d *Daemon) error
 	return nil
 }
 
+func patchSplitIdentityCertificateEntityTypes(_ string, d *Daemon) error {
+	s := d.State()
+	err := s.DB.Cluster.Transaction(s.ShutdownCtx, func(ctx context.Context, tx *db.ClusterTx) error {
+		// Notes:
+		// - We don't need to handle warnings, there are no warnings for identities.
+		// - No permissions have been created for OIDC identities against the certificate entity type, because the database
+		//   definition for the certificate entity type ensures that the authentication method is TLS.
+		// - We only need to fix permissions defined with the "identity" entity type that are certificates.
+
+		// Get permissions applied to the identity entity type, that really are certificates (auth method = TLS)
+		incorrectIdentityPermissionIDs, err := query.SelectIntegers(ctx, tx.Tx(), `
+SELECT auth_groups_permissions.id FROM auth_groups_permissions
+	JOIN identities ON auth_groups_permissions.entity_id = identities.id
+	WHERE auth_groups_permissions.entity_type = 24
+	AND identities.auth_method = 1
+`)
+		if err != nil {
+			return err
+		}
+
+		args := make([]any, 0, len(incorrectIdentityPermissionIDs))
+		for _, id := range incorrectIdentityPermissionIDs {
+			args = append(args, id)
+		}
+
+		// Change these permissions to apply to certificates instead. Use "UPDATE OR REPLACE" in case of UNIQUE constraint violation.
+		stmt := fmt.Sprintf("UPDATE OR REPLACE auth_groups_permissions SET entity_type = 4 WHERE id IN %s", query.Params(len(args)))
```

Comment:
Please can you define the magic numbers 24, 1, and 4 as local vars/constants in this function for future readability?

Reply:
Yep. Those numbers are the entity type and authentication method constants that are intentionally not exported from the `lxd/db/cluster` package because we should be using constants from `shared/api` or `shared/entity` everywhere else.

Reply:
Rather than define the magic numbers here, I've used the `Value` method on the database type definitions for these. E.g.
```
certificateEntityTypeCode, _ := dbCluster.EntityType(entity.TypeCertificate).Value()
```